src/DynamoPackages/PackageManagerExtension.cs

```diff
@@ -149,6 +176,87 @@ private void OnMessageLogged(ILogMessage msg)
                 this.MessageLogged(msg);
             }
         }
+        
+        private void OnCurrentWorkspaceChanged(IWorkspaceModel ws)
+        {
+            if (ws is WorkspaceModel)
+            {
+                if (currentWorkspace != null)
+                {
+                    (currentWorkspace as WorkspaceModel).CollectingCustomNodePackageDependencies -= GetCustomNodePackageFromID;
+                    (currentWorkspace as WorkspaceModel).CollectingNodePackageDependencies -= GetNodePackageFromAssemblyName;
+                }
+                
+                (ws as WorkspaceModel).CollectingCustomNodePackageDependencies += GetCustomNodePackageFromID;
+                (ws as WorkspaceModel).CollectingNodePackageDependencies += GetNodePackageFromAssemblyName;
+                currentWorkspace = ws;
+            }
+        }
+
+        private PackageInfo GetNodePackageFromAssemblyName(AssemblyName assemblyName)
+        {
+            if (NodePackageDictionary.ContainsKey(assemblyName.FullName))
+            {
+                return NodePackageDictionary[assemblyName.FullName];
+            }
+            return null;
+        }
+
+        private PackageInfo GetCustomNodePackageFromID(Guid functionID)
+        {
+            if (CustomNodePackageDictionary.ContainsKey(functionID))
+            {
+                return CustomNodePackageDictionary[functionID];
+            }
+            return null;
+        }
+
+        private void OnPackageLoaded(Package package)
+        {
+            // Create NodePackageDictionary if it doesn't exist
+            if (NodePackageDictionary == null)
+            {
+                NodePackageDictionary = new Dictionary<string, PackageInfo>();
+            }
+            // Add new assemblies to NodePackageDictionary
+            foreach (var assembly in package.LoadedAssemblies.Select(a => AssemblyName.GetAssemblyName(a.Assembly.Location)))
+            {
+                NodePackageDictionary[assembly.FullName] = new PackageInfo(package.Name, package.VersionName);
```

Comment:
so what happens if somehow theres already a mapped package here at this key - I think we should log something to console or throw - this could happen if two packages reference the same .dll and expose nodes from it... I think this is related to what @aparajit-pratap is working on. So perhaps this needs to be `lists of <PackageInfo>`